File: app/src/main/java/net/j4dy/familypicingame/ui/screens/FaceCropScreen.kt
package net.j4dy.familypicingame.ui.screens

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.net.Uri
import androidx.activity.compose.BackHandler
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.gestures.detectTransformGestures
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ArrowBack
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.*
import androidx.compose.ui.graphics.drawscope.clipPath
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.layout.onSizeChanged
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.IntOffset
import androidx.compose.ui.unit.IntSize
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import net.j4dy.familypicingame.data.FaceStorage
import net.j4dy.familypicingame.ui.theme.CyberPurple
import net.j4dy.familypicingame.ui.theme.ElectricCyan
import net.j4dy.familypicingame.ui.theme.IcyWhite
import net.j4dy.familypicingame.ui.theme.NeonPink
import net.j4dy.familypicingame.ui.theme.SoftGrey
import java.io.InputStream

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun FaceCropScreen(
    imageUri: Uri,
    onCropSuccess: () -> Unit,
    onBackClick: () -> Unit
) {
    val context = LocalContext.current
    val faceStorage = remember { FaceStorage(context) }
    
    // Load bitmap in background safely
    var bitmap by remember { mutableStateOf<Bitmap?>(null) }
    var loadingError by remember { mutableStateOf(false) }
    
    LaunchedEffect(imageUri) {
        try {
            bitmap = loadDownsampledBitmap(context, imageUri, 1024)
        } catch (e: Exception) {
            e.printStackTrace()
            loadingError = true
        }
    }
    
    var nameText by remember { mutableStateOf("") }
    var showError by remember { mutableStateOf(false) }
    
    // Transform states for cropping
    var scale by remember { mutableStateOf(1f) }
    var offset by remember { mutableStateOf(Offset.Zero) }
    var canvasSize by remember { mutableStateOf(Size.Zero) }

    BackHandler {
        onBackClick()
    }

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Crop Character Head", fontWeight = FontWeight.Bold) },
                navigationIcon = {
                    IconButton(onClick = onBackClick) {
                        Icon(Icons.Default.ArrowBack, contentDescription = "Back", tint = IcyWhite)
                    }
                },
                colors = TopAppBarDefaults.topAppBarColors(
                    containerColor = MaterialTheme.colorScheme.background,
                    titleContentColor = IcyWhite
                )
            )
        }
    ) { padding ->
        if (loadingError) {
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(padding)
                    .background(MaterialTheme.colorScheme.background),
                contentAlignment = Alignment.Center
            ) {
                Text("Failed to load image. Please try another.", color = NeonPink, textAlign = TextAlign.Center)
            }
        } else if (bitmap == null) {
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(padding)
                    .background(MaterialTheme.colorScheme.background),
                contentAlignment = Alignment.Center
            ) {
                CircularProgressIndicator(color = NeonPink)
            }
        } else {
            Column(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(padding)
                    .background(MaterialTheme.colorScheme.background)
                    .padding(16.dp),
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.SpaceBetween
            ) {
                Text(
                    "Pinch to zoom and drag to position family member's head inside the circle.",
                    style = MaterialTheme.typography.bodyMedium,
                    textAlign = TextAlign.Center,
                    modifier = Modifier.padding(bottom = 12.dp)
                )

                // The Crop Canvas
                Box(
                    modifier = Modifier
                        .weight(1f)
                        .fillMaxWidth()
                        .clip(RoundedCornerShape(16.dp))
                        .background(Color(0xFF0F1424))
                        .onSizeChanged { intSize ->
                            canvasSize = Size(intSize.width.toFloat(), intSize.height.toFloat())
                        }
                        .pointerInput(Unit) {
                            detectTransformGestures { _, panDrag, zoomMultiplier, _ ->
                                scale = (scale * zoomMultiplier).coerceIn(0.5f, 5.0f)
                                offset = offset + panDrag
                            }
                        },
                    contentAlignment = Alignment.Center
                ) {
                    val imageBitmap = remember(bitmap) { bitmap?.asImageBitmap() }
                    
                    if (imageBitmap != null) {
                        Canvas(modifier = Modifier.fillMaxSize()) {
                            val canvasWidth = size.width
                            val canvasHeight = size.height
                            val circleRadius = Math.min(canvasWidth, canvasHeight) * 0.35f
                            val center = Offset(canvasWidth / 2f, canvasHeight / 2f)

                            // 1. Draw image with scale and offset centered
                            val imgWidth = imageBitmap.width.toFloat()
                            val imgHeight = imageBitmap.height.toFloat()
                            
                            val baseScale = Math.min(canvasWidth / imgWidth, canvasHeight / imgHeight)
                            val finalScale = baseScale * scale
                            
                            val drawWidth = imgWidth * finalScale
                            val drawHeight = imgHeight * finalScale
                            
                            val startX = (canvasWidth - drawWidth) / 2f + offset.x
                            val startY = (canvasHeight - drawHeight) / 2f + offset.y

                            // Draw image
                            drawImage(
                                image = imageBitmap,
                                dstOffset = IntOffset(startX.toInt(), startY.toInt()),
                                dstSize = IntSize(drawWidth.toInt(), drawHeight.toInt())
                            )

                            // 2. Draw glass overlay with circular hole
                            val path = Path().apply {
                                addRect(Rect(0f, 0f, canvasWidth, canvasHeight))
                            }
                            val circlePath = Path().apply {
                                addOval(Rect(center.x - circleRadius, center.y - circleRadius, center.x + circleRadius, center.y + circleRadius))
                            }
                            
                            // Subtract circle from rect
                            val differencePath = Path.combine(
                                PathOperation.Difference,
                                path,
                                circlePath
                            )
                            
                            drawPath(
                                path = differencePath,
                                color = Color(0xAA0A0E17)
                            )
                            
                            // Draw neon border circle helper
                            drawCircle(
                                color = ElectricCyan,
                                radius = circleRadius,
                                center = center,
                                style = androidx.compose.ui.graphics.drawscope.Stroke(width = 3.dp.toPx())
                            )
                        }
                    }
                }

                Spacer(modifier = Modifier.height(16.dp))

                // Name Input field
                OutlinedTextField(
                    value = nameText,
                    onValueChange = {
                        nameText = it
                        showError = false
                    },
                    label = { Text("Family Member Name (e.g., Dad, Mom)") },
                    singleLine = true,
                    isError = showError,
                    modifier = Modifier.fillMaxWidth(),
                    colors = OutlinedTextFieldDefaults.colors(
                        focusedBorderColor = ElectricCyan,
                        unfocusedBorderColor = CyberPurple,
                        focusedLabelColor = ElectricCyan,
                        unfocusedLabelColor = SoftGrey,
                        focusedTextColor = IcyWhite,
                        unfocusedTextColor = IcyWhite
                    ),
                    shape = RoundedCornerShape(12.dp)
                )

                if (showError) {
                    Text(
                        "Please enter a valid name.",
                        color = NeonPink,
                        style = MaterialTheme.typography.bodySmall,
                        modifier = Modifier.align(Alignment.Start).padding(start = 8.dp, top = 4.dp)
                    )
                }

                Spacer(modifier = Modifier.height(16.dp))

                // Save button
                Button(
                    onClick = {
                        if (nameText.trim().isEmpty()) {
                            showError = true
                        } else {
                            // Extract crop and save
                            val b = bitmap
                            if (b != null) {
                                val cropped = cropBitmapNatively(b, scale, offset, canvasSize.width, canvasSize.height)
                                if (cropped != null) {
                                    faceStorage.addProfile(nameText.trim(), cropped)
                                    onCropSuccess()
                                }
                            }
                        }
                    },
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(56.dp),
                    colors = ButtonDefaults.buttonColors(
                        containerColor = NeonPink,
                        contentColor = Color.White
                    ),
                    shape = RoundedCornerShape(12.dp)
                ) {
                    Text("Save Character Face", fontSize = 18.sp, fontWeight = FontWeight.Bold)
                }
            }
        }
    }
}

/**
 * Perform actual crop using Matrix transformations to extract precisely what the user aligned.
 */
private fun cropBitmapNatively(
    src: Bitmap,
    gestureScale: Float,
    gestureOffset: Offset,
    canvasWidth: Float,
    canvasHeight: Float
): Bitmap? {
    try {
        // A standard crop dimension (e.g. 256x256 makes a perfect avatar)
        val size = 256
        val croppedOutput = Bitmap.createBitmap(size, size, Bitmap.Config.ARGB_8888)
        val canvas = android.graphics.Canvas(croppedOutput)
        
        // Paint setup with anti-aliasing and bitmap filtering for smooth crops
        val paint = android.graphics.Paint().apply { 
            isAntiAlias = true 
            isFilterBitmap = true
        }

        val widthVal = if (canvasWidth > 0f) canvasWidth else 800f
        val heightVal = if (canvasHeight > 0f) canvasHeight else 800f

        val circleRadius = Math.min(widthVal, heightVal) * 0.35f
        val centerX = widthVal / 2f
        val centerY = heightVal / 2f

        val imgWidth = src.width.toFloat()
        val imgHeight = src.height.toFloat()

        // Replicate base scale and final scale from preview
        val baseScale = Math.min(widthVal / imgWidth, heightVal / imgHeight)
        val finalScale = baseScale * gestureScale

        val drawWidth = imgWidth * finalScale
        val drawHeight = imgHeight * finalScale

        // Replicate initial centering start coords
        val startX = (widthVal - drawWidth) / 2f + gestureOffset.x
        val startY = (heightVal - drawHeight) / 2f + gestureOffset.y

        val transform = Matrix()
        // 1. Position and scale image exactly as visually laid out on physical screen
        transform.postScale(finalScale, finalScale)
        transform.postTranslate(startX, startY)

        // 2. Shift center of physical screen's crop circle to (0, 0)
        transform.postTranslate(-centerX, -centerY)

        // 3. Scale crop circle radius to output bitmap's radius (128f)
        val cropScale = 128f / circleRadius
        transform.postScale(cropScale, cropScale)

        // 4. Translate center (0, 0) to output bitmap's center (128f, 128f)
        transform.postTranslate(128f, 128f)

        // Clean slate and clear
        canvas.drawColor(android.graphics.Color.TRANSPARENT, android.graphics.PorterDuff.Mode.CLEAR)
        
        // Clip to circular crop path so everything outside the 128px radius circle is transparent
        val clipPath = android.graphics.Path().apply {
            addCircle(128f, 128f, 128f, android.graphics.Path.Direction.CW)
        }
        canvas.clipPath(clipPath)
        
        canvas.drawBitmap(src, transform, paint)
        
        return croppedOutput
    } catch (e: Exception) {
        e.printStackTrace()
    }
    return null
}

/**
 * Loads and downsamples a bitmap from a Uri to prevent memory crashes.
 */
private fun loadDownsampledBitmap(context: Context, uri: Uri, maxDim: Int): Bitmap? {
    var inputStream: InputStream? = null
    try {
        inputStream = context.contentResolver.openInputStream(uri)
        val options = BitmapFactory.Options().apply {
            inJustDecodeBounds = true
        }
        BitmapFactory.decodeStream(inputStream, null, options)
        inputStream?.close()

        var width = options.outWidth
        var height = options.outHeight
        var sampleSize = 1

        while (width > maxDim || height > maxDim) {
            width /= 2
            height /= 2
            sampleSize *= 2
        }

        val decodeOptions = BitmapFactory.Options().apply {
            inSampleSize = sampleSize
        }
        inputStream = context.contentResolver.openInputStream(uri)
        val result = BitmapFactory.decodeStream(inputStream, null, decodeOptions)
        
        // Handle rotation if needed (panned images are sometimes rotated in EXIF)
        return result
    } catch (e: Exception) {
        e.printStackTrace()
    } finally {
        inputStream?.close()
    }
    return null
}

File: lib/screens/face_crop_screen.dart
import 'dart:io';
import 'dart:math' as math;
import 'dart:ui' as ui;
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import '../../theme.dart';
import '../../data/face_storage.dart';

class FaceCropScreen extends StatefulWidget {
  final File imageFile;
  final FaceStorage faceStorage;
  final VoidCallback onCropSuccess;
  final VoidCallback onBackClick;

  const FaceCropScreen({
    super.key,
    required this.imageFile,
    required this.faceStorage,
    required this.onCropSuccess,
    required this.onBackClick,
  });

  @override
  State<FaceCropScreen> createState() => _FaceCropScreenState();
}

class _HomeScreenState {}

class _FaceCropScreenState extends State<FaceCropScreen> {
  ui.Image? _uiImage;
  bool _loadingError = false;
  bool _isSaving = false;

  final TextEditingController _nameController = TextEditingController();
  bool _showError = false;

  // Touch transform states
  double _scale = 1.0;
  double _baseScale = 1.0;
  Offset _offset = Offset.zero;
  Offset _baseOffset = Offset.zero;

  Size _canvasSize = Size.zero;

  @override
  void initState() {
    super.initState();
    _loadImage();
  }

  Future<void> _loadImage() async {
    try {
      final bytes = await widget.imageFile.readAsBytes();
      // Downsample image to max width of 1024 to prevent memory issues
      final codec = await ui.instantiateImageCodec(bytes, targetWidth: 1024);
      final frame = await codec.getNextFrame();
      setState(() {
        _uiImage = frame.image;
      });
    } catch (e) {
      debugPrint('Error loading image: $e');
      setState(() {
        _loadingError = true;
      });
    }
  }

  @override
  void dispose() {
    _nameController.dispose();
    super.dispose();
  }

  Future<void> _saveCrop() async {
    final name = _nameController.text.trim();
    if (name.isEmpty) {
      setState(() {
        _showError = true;
      });
      return;
    }

    final img = _uiImage;
    if (img == null || _canvasSize == Size.zero) return;

    setState(() {
      _isSaving = true;
    });

    try {
      final croppedImage = await _cropImageNatively(
        img,
        _scale,
        _offset,
        _canvasSize.width,
        _canvasSize.height,
      );

      await widget.faceStorage.addProfile(name, croppedImage);
      widget.onCropSuccess();
    } catch (e) {
      debugPrint('Error cropping image: $e');
      ScaffoldMessenger.of(context).showSnackBar(
        SnackBar(
          content: Text('Error saving cropped head: $e'),
          backgroundColor: neonPink,
        ),
      );
    } finally {
      setState(() {
        _isSaving = false;
      });
    }
  }

  Future<ui.Image> _cropImageNatively(
    ui.Image src,
    double gestureScale,
    Offset gestureOffset,
    double canvasWidth,
    double canvasHeight,
  ) async {
    final recorder = ui.PictureRecorder();
    final canvas = Canvas(recorder);

    const double size = 256.0;
    final double widthVal = canvasWidth > 0 ? canvasWidth : 800.0;
    final double heightVal = canvasHeight > 0 ? canvasHeight : 800.0;

    final double circleRadius = math.min(widthVal, heightVal) * 0.35;
    final double centerX = widthVal / 2.0;
    final double centerY = heightVal / 2.0;

    final double imgWidth = src.width.toDouble();
    final double imgHeight = src.height.toDouble();

    // Replicate layout scale calculations
    final double baseScale = math.min(widthVal / imgWidth, heightVal / imgHeight);
    final double finalScale = baseScale * gestureScale;

    final double drawWidth = imgWidth * finalScale;
    final double drawHeight = imgHeight * finalScale;

    // Center coordinates
    final double startX = (widthVal - drawWidth) / 2.0 + gestureOffset.dx;
    final double startY = (heightVal - drawHeight) / 2.0 + gestureOffset.dy;

    // 1. Clip to a circular crop path in 256x256 dimensions
    final clipPath = Path()..addOval(const Rect.fromLTRB(0, 0, size, size));
    canvas.clipPath(clipPath);

    // 2. Perform matrix transform matching preview layout
    canvas.translate(128.0, 128.0);
    canvas.scale(128.0 / circleRadius);
    canvas.translate(startX - centerX, startY - centerY);
    canvas.scale(finalScale);

    final paint = Paint()
      ..isAntiAlias = true
      ..filterQuality = FilterQuality.high;

    canvas.drawImage(src, Offset.zero, paint);

    final picture = recorder.endRecording();
    return await picture.toImage(256, 256);
  }

  @override
  Widget build(BuildContext context) {
    if (_loadingError) {
      return Scaffold(
        appBar: AppBar(title: const Text("Crop Character Head")),
        body: const Center(
          child: Text(
            "Failed to load image. Please try another.",
            style: TextStyle(color: neonPink),
            textAlign: TextAlign.center,
          ),
        ),
      );
    }

    if (_uiImage == null) {
      return Scaffold(
        appBar: AppBar(title: const Text("Crop Character Head")),
        body: const Center(
          child: CircularProgressIndicator(color: neonPink),
        ),
      );
    }

    return Scaffold(
      appBar: AppBar(
        title: const Text("Crop Character Head"),
        leading: IconButton(
          icon: const Icon(Icons.arrow_back),
          onPressed: widget.onBackClick,
        ),
      ),
      body: SafeArea(
        child: Padding(
          padding: const EdgeInsets.all(16.0),
          child: Column(
            children: [
              const Text(
                "Pinch to zoom and drag to position family member's head inside the circle.",
                style: TextStyle(color: softGrey, fontSize: 14),
                textAlign: TextAlign.center,
              ),
              const SizedBox(height: 12),

              // The Crop Canvas
              Expanded(
                child: ClipRRect(
                  borderRadius: BorderRadius.circular(16),
                  child: Container(
                    color: const Color(0xFF0F1424),
                    child: LayoutBuilder(
                      builder: (context, constraints) {
                        WidgetsBinding.instance.addPostFrameCallback((_) {
                          final size = Size(constraints.maxWidth, constraints.maxHeight);
                          if (_canvasSize != size) {
                            setState(() {
                              _canvasSize = size;
                            });
                          }
                        });

                        return GestureDetector(
                          onScaleStart: (details) {
                            _baseScale = _scale;
                            _baseOffset = _offset;
                          },
                          onScaleUpdate: (details) {
                            setState(() {
                              _scale = (_baseScale * details.scale).clamp(0.5, 5.0);
                              _offset = _baseOffset + details.focalPointDelta;
                            });
                          },
                          child: CustomPaint(
                            size: Size.infinite,
                            painter: _CropPreviewPainter(
                              image: _uiImage!,
                              scale: _scale,
                              offset: _offset,
                            ),
                          ),
                        );
                      },
                    ),
                  ),
                ),
              ),

              const SizedBox(height: 16),

              // Name Input field
              TextField(
                controller: _nameController,
                onChanged: (text) {
                  if (_showError) {
                    setState(() {
                      _showError = false;
                    });
                  }
                },
                style: const TextStyle(color: icyWhite),
                decoration: InputDecoration(
                  labelText: "Family Member Name (e.g., Dad, Mom)",
                  labelStyle: const TextStyle(color: softGrey),
                  focusedBorder: OutlineInputBorder(
                    borderRadius: BorderRadius.circular(12),
                    borderSide: const BorderSide(color: electricCyan),
                  ),
                  enabledBorder: OutlineInputBorder(
                    borderRadius: BorderRadius.circular(12),
                    borderSide: const BorderSide(color: cyberPurple),
                  ),
                  errorText: _showError ? "Please enter a valid name." : null,
                  errorStyle: const TextStyle(color: neonPink),
                ),
              ),

              const SizedBox(height: 16),

              // Save button
              ElevatedButton(
                onPressed: _isSaving ? null : _saveCrop,
                style: ElevatedButton.styleFrom(
                  backgroundColor: neonPink,
                  minimumSize: const Size.fromHeight(56),
                  shape: RoundedRectangleBorder(
                    borderRadius: BorderRadius.circular(12),
                  ),
                ),
                child: _isSaving
                    ? const CircularProgressIndicator(color: Colors.white)
                    : const Text(
                        "SAVE CHARACTER FACE",
                        style: TextStyle(
                          fontSize: 16,
                          fontWeight: FontWeight.bold,
                          color: Colors.white,
                        ),
                      ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

class _CropPreviewPainter extends CustomPainter {
  final ui.Image image;
  final double scale;
  final Offset offset;

  _CropPreviewPainter({
    required this.image,
    required this.scale,
    required this.offset,
  });

  @override
  void paint(Canvas canvas, Size size) {
    final canvasWidth = size.width;
    final canvasHeight = size.height;
    final circleRadius = math.min(canvasWidth, canvasHeight) * 0.35;
    final center = Offset(canvasWidth / 2.0, canvasHeight / 2.0);

    // 1. Draw image with scale and offset centered
    final imgWidth = image.width.toDouble();
    final imgHeight = image.height.toDouble();

    final baseScale = math.min(canvasWidth / imgWidth, canvasHeight / imgHeight);
    final finalScale = baseScale * scale;

    final drawWidth = imgWidth * finalScale;
    final drawHeight = imgHeight * finalScale;

    final startX = (canvasWidth - drawWidth) / 2.0 + offset.dx;
    final startY = (canvasHeight - drawHeight) / 2.0 + offset.dy;

    canvas.save();
    canvas.translate(startX, startY);
    canvas.scale(finalScale);

    final paint = Paint()
      ..isAntiAlias = true
      ..filterQuality = FilterQuality.high;
    canvas.drawImage(image, Offset.zero, paint);
    canvas.restore();

    // 2. Draw glass overlay with circular hole
    final path = Path()..addRect(Rect.fromLTRB(0, 0, canvasWidth, canvasHeight));
    final circlePath = Path()
      ..addOval(Rect.fromCircle(center: center, radius: circleRadius));

    // Subtract circle from rect (Difference)
    final differencePath = Path.combine(
      PathOperation.difference,
      path,
      circlePath,
    );

    final overlayPaint = Paint()
      ..color = const Color(0xAA0A0E17)
      ..style = PaintingStyle.fill;
    canvas.drawPath(differencePath, overlayPaint);

    // Draw neon border circle helper
    final borderPaint = Paint()
      ..color = electricCyan
      ..style = PaintingStyle.stroke
      ..strokeWidth = 3.0;
    canvas.drawCircle(center, circleRadius, borderPaint);
  }

  @override
  bool shouldRepaint(covariant _CropPreviewPainter oldDelegate) {
    return oldDelegate.image != image ||
        oldDelegate.scale != scale ||
        oldDelegate.offset != offset;
  }
}
